File: sliding_window/longest_repeating_character_replacement.py
import sys

class Solution:
    # brute-force method
    def characterReplacement(self, s: str, k: int) -> int:
        count = {}
        longestLength = l = r = 0

        while r < len(s):
            count[s[r]] = count.get(s[r], 0) + 1

            # O(26 * n)
            highestFrequency = max(count.values())

            while (r - l - highestFrequency + 1 > k):
                count[s[l]] -= 1
                l += 1

            longestLength = max(longestLength, r - l + 1)

            r += 1

        return longestLength

    def optimisedSolution(self, s: str, k: int) -> int:
        count = {}
        highestFrequency = longestLength = l = r = 0

        while r < len(s):
            count[s[r]] = count.get(s[r], 0) + 1

            highestFrequency = max(count[s[r]], highestFrequency)

            while (r - l - highestFrequency + 1 > k):
                count[s[l]] -= 1
                l += 1

            longestLength = max(longestLength, r - l + 1)

            r += 1

        return longestLength

if __name__ == "__main__":
    s = Solution()
    print(s.characterReplacement(sys.argv[1], int(sys.argv[2])))
    print(s.optimisedSolution(sys.argv[1], int(sys.argv[2])))

File: sliding_window/longest_repeating_character_replacement.cpp
class Solution {
public:
  int characterReplacement(string s, int k) {
    int chars[26] = {};
    int longest_length = 0;
    int substring_start = 0;
    for (int j = 0; j < s.length(); j++) {
      chars[s[j] - 65] += 1;
      // using a while loop here provides no benefits as we can not care
      // about an invalid window that is of the same size as the max length
      // we have currently found. the window size will only extend if we find
      // another valid window that is longer than the current one!
      if (j - substring_start + 1 - *std::max_element(chars, chars + 26) > k) {
        chars[s[substring_start] - 65] -= 1;
        substring_start += 1;
      }
      longest_length = std::max(longest_length, j - substring_start + 1);
    }
    return longest_length;
  }
};
